### PR TITLE
Make NPC class id a string_id

### PR DESCRIPTION
**Pull request by Coolthulhu**
The hardcoded `enum` NPC class is one of the many things that make NPC code bad and it encourages horrible code full of switches. It also makes it much harder to properly jsonize anything NPC-related.

Here I introduce a `npc_class` class that uses `string_id` and `generic_factory`.

Currently the class is rather bare, since I didn't want to bloat the PR further.

I also removed an unused field and enum from NPCs. The `flags` field was supposed to give NPCs some extra personality and different priorities, but its effects were either never there or lost in development.
They could be reimplemented once NPC classes are jsonized, but not in the crappy form they have at the moment (hardcoded and using bit operations on `int`s).

### CODE REVIEW

File: src/npc_class.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
+#include "npc_class.h"
+#include "debug.h"
+#include "rng.h"
+#include "generic_factory.h"
+
+#include <list>
+
+static const std::array<npc_class_id, 17> legacy_ids = {{
+        npc_class_id( "NC_NONE" ),
```
Comment (codemime) [inline src/npc_class.cpp]: You could probably move the array down below and write there:
```c++
static const std::array<npc_class_id, 17> legacy_ids = {{
    NC_NONE,
    NC_EVAC_SHOPKEEP,
    ...
```

File: src/npc_class.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        npc_class_id( "NC_SCAVENGER" ),      // Good with pistols light weapons
+        npc_class_id( "NC_ARSONIST" ),       // Evacuation Center, restocks moltovs and anarcist type stuff
+        npc_class_id( "NC_HUNTER" ),         // Survivor type good with bow or rifle
+        npc_class_id( "NC_SOLDIER" ),        // Well equiped and trained combatant, good with rifles and melee
+        npc_class_id( "NC_BARTENDER" ),      // Stocks alcohol
+        npc_class_id( "NC_JUNK_SHOPKEEP" )   // Stocks wide range of items...
+    }
+};
+
+npc_class_id NC_NONE( "NC_NONE" );
```
Comment (codemime) [inline src/npc_class.cpp]: Shouldn't they be `static const`'s?

Comment (BevapDin) [inline src/npc_class.cpp]: They are `extern` and can therefor not be `static`, but they should indeed be `const`. But maybe the current for is intended to encourage people rid of the mas soon as possible (-:

Comment (Coolthulhu) [inline src/npc_class.cpp]: Yes, I mostly left them only for compatibility. A lot of code that should be in JSONs is instead in `if( class_id == NC_WHATEVER )` blocks.

**Pull request merged**